## First: Review Existing Unstructured Data and Diagram a New Structured Relational Data Model

![title](img/Untitled.PNG )

## Second: Write a query that directly answers a predetermined question from a business stakeholder

### Importing Libraries 

In [21]:
import gzip
import json
import pandas as pd
import datetime

###  Reading Data 

### Receipts

In [22]:
with gzip.open("receipts.json.gz", "r") as f:
    data_receipts = [json.loads(line) for line in f]

receipt_id=[]
for a in range(len(data_receipts)):
    if '_id' in data_receipts[a]:
        receipt_id.append( data_receipts[a]['_id']['$oid'] )
    else:
        receipt_id.append('Null')

        
bonusPointsEarned=[]
for a in range(len(data_receipts)):
    if 'bonusPointsEarned' in data_receipts[a]:
        bonusPointsEarned.append( data_receipts[a]['bonusPointsEarned'] )
    else:
        bonusPointsEarned.append('Null')

bonusPointsEarnedReason=[]
for a in range(len(data_receipts)):
    if 'bonusPointsEarnedReason' in data_receipts[a]:
        bonusPointsEarnedReason.append( data_receipts[a]['bonusPointsEarnedReason'] )
    else:
        bonusPointsEarnedReason.append('Null')
           
createDate=[]
for a in range(len(data_receipts)):
    if 'createDate' in data_receipts[a]:
        createDate.append(  pd.to_datetime(data_receipts[a]['createDate']['$date'] , unit='ms'))
    else:
        createDate.append('Null')
        
        
dateScanned=[]
for a in range(len(data_receipts)):
    if 'dateScanned' in data_receipts[a]:
        dateScanned.append(   pd.to_datetime(data_receipts[a]['dateScanned']['$date'] , unit='ms'))
    else:
        dateScanned.append('Null')
        
        
finishedDate=[]
for a in range(len(data_receipts)):
    if 'finishedDate' in data_receipts[a]:
        finishedDate.append(  pd.to_datetime(data_receipts[a]['finishedDate']['$date'] , unit='ms'))
    else:
        finishedDate.append('Null')
        
        
modifyDate=[]
for a in range(len(data_receipts)):
    if 'modifyDate' in data_receipts[a]:
        modifyDate.append(  pd.to_datetime(data_receipts[a]['modifyDate']['$date'] , unit='ms'))
    else:
        modifyDate.append('Null')
        
pointsAwardedDate=[]
for a in range(len(data_receipts)):
    if 'pointsAwardedDate' in data_receipts[a]:
        pointsAwardedDate.append(  pd.to_datetime(data_receipts[a]['pointsAwardedDate']['$date'] , unit='ms'))
    else:
        pointsAwardedDate.append('Null')
        
pointsEarned=[]
for a in range(len(data_receipts)):
    if 'pointsEarned' in data_receipts[a]:
        pointsEarned.append( data_receipts[a]['pointsEarned']) 
    else:
        pointsEarned.append('Null')
        
purchaseDate=[]
for a in range(len(data_receipts)):
    if 'purchaseDate' in data_receipts[a]:
        purchaseDate.append(   pd.to_datetime(data_receipts[a]['purchaseDate']['$date'] , unit='ms'))
    else:
        purchaseDate.append('Null')
        
purchasedItemCount=[]
for a in range(len(data_receipts)):
    if 'purchasedItemCount' in data_receipts[a]:
        purchasedItemCount.append( data_receipts[a]['purchasedItemCount'] )
    else:
        purchasedItemCount.append('Null')
        
             
rewardsReceiptStatus=[]
for a in range(len(data_receipts)):
    if 'rewardsReceiptStatus' in data_receipts[a]:
        rewardsReceiptStatus.append( data_receipts[a]['rewardsReceiptStatus'] )
    else:
        rewardsReceiptStatus.append('Null')
        
totalSpent=[]
for a in range(len(data_receipts)):
    if 'totalSpent' in data_receipts[a]:
        totalSpent.append( data_receipts[a]['totalSpent'] )
    else:
        totalSpent.append('Null')
        
userId=[]
for a in range(len(data_receipts)):
    if 'userId' in data_receipts[a]:
        userId.append( data_receipts[a]['userId'] )
    else:
        userId.append('Null')
        
######        
receipt_id_description_rewardsGroup=[]

for a in range(len(data_receipts)):
         
    if 'rewardsReceiptItemList' in data_receipts[a]:
        for b in range(len(data_receipts[a]['rewardsReceiptItemList'])):
            new=[]
            if '_id' in data_receipts[a]:
                new.append( data_receipts[a]['_id']['$oid'] )
            else:
                new.append('Null')
            if 'description' in data_receipts[a]['rewardsReceiptItemList'][b]:
                new.append(  data_receipts[a]['rewardsReceiptItemList'][b]['description']  )
            else:
                new.append('Null')
                
                
            if 'rewardsGroup' in data_receipts[a]['rewardsReceiptItemList'][b]:
                new.append(  data_receipts[a]['rewardsReceiptItemList'][b]['rewardsGroup']  )
            else:
                new.append('Null')
            
            receipt_id_description_rewardsGroup.append(new)
                
    else:
        new=[]
        if '_id' in data_receipts[a]:
            new.append( data_receipts[a]['_id']['$oid'] )
        else:
            new.append('Null')
        new.append('Null')
        new.append('Null')
        
        receipt_id_description_rewardsGroup.append(new)
#####

df_receipts_1 = pd.DataFrame({'receipt_id':receipt_id, 'bonusPointsEarned':bonusPointsEarned, 'bonusPointsEarnedReason':bonusPointsEarnedReason,
            'createDate':createDate, 'dateScanned':dateScanned, 'finishedDate':finishedDate,
            'modifyDate':modifyDate, 'pointsAwardedDate':pointsAwardedDate, 'pointsEarned':pointsEarned,
             'purchaseDate':purchaseDate, 'purchasedItemCount': purchasedItemCount, 
              'rewardsReceiptStatus': rewardsReceiptStatus,
             'totalSpent': totalSpent, 'userId': userId})


df_receipts_2 = pd.DataFrame(data = receipt_id_description_rewardsGroup, columns=['receipt_id', 'description','rewardsGroup'])

df_receipts = pd.merge(df_receipts_2, df_receipts_1 , on = 'receipt_id' ,how= 'left')

def func_receipt(s):
    if s=='Null':
        return 'Null'
    else:
        return s.split()[0]
df_receipts['rewardsGroup_cleaned']=df_receipts['rewardsGroup'].apply(func_receipt)

df_receipts.head()

receipt_id  \
0  5ff1e1eb0a720f0523000575   
1  5ff1e1bb0a720f052300056b   
2  5ff1e1bb0a720f052300056b   
3  5ff1e1f10a720f052300057a   
4  5ff1e1ee0a7214ada100056f   

                                         description  \
0                                     ITEM NOT FOUND   
1                                     ITEM NOT FOUND   
2  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...   
3                                               Null   
4                                     ITEM NOT FOUND   

                             rewardsGroup bonusPointsEarned  \
0                                    Null               500   
1                                    Null               150   
2  DORITOS SPICY SWEET CHILI SINGLE SERVE               150   
3                                    Null                 5   
4                                    Null                 5   

                             bonusPointsEarnedReason          createDate  \
0  Receipt number 2 completed, bonus point schedu... 2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
2  Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
3                         All-receipts receipt bonus 2021-01-03 15:25:37   
4                         All-receipts receipt bonus 2021-01-03 15:25:34   

          dateScanned         finishedDate          modifyDate  \
0 2021-01-03 15:25:31  2021-01-03 15:25:31 2021-01-03 15:25:36   
1 2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
2 2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
3 2021-01-03 15:25:37                 Null 2021-01-03 15:25:42   
4 2021-01-03 15:25:34  2021-01-03 15:25:34 2021-01-03 15:25:39   

     pointsAwardedDate pointsEarned         purchaseDate purchasedItemCount  \
0  2021-01-03 15:25:31        500.0  2021-01-03 00:00:00                  5   
1  2021-01-03 15:24:43        150.0  2021-01-02 15:24:43                  2   
2  2021-01-03 15:24:43        150.0  2021-01-02 15:24:43                  2   
3                 Null            5  2021-01-03 00:00:00                  1   
4  2021-01-03 15:25:34          5.0  2021-01-03 00:00:00                  4   

  rewardsReceiptStatus totalSpent                    userId  \
0             FINISHED      26.00  5ff1e1eacfcf6c399c274ae6   
1             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052   
2             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052   
3             REJECTED      10.00  5ff1e1f1cfcf6c399c274b0b   
4             FINISHED      28.00  5ff1e1eacfcf6c399c274ae6   

  rewardsGroup_cleaned  
0                 Null  
1                 Null  
2              DORITOS  
3                 Null  
4                 Null

In [23]:
from sqlalchemy import create_engine

engine1 = create_engine('sqlite://', echo = False)

df_receipts = df_receipts.applymap(str)

df_receipts.to_sql('new1', con = engine1)

K= engine1.execute("SELECT * FROM new1").fetchall()
K= [a[1:] for a in K]

In [1]:
for row in K:
    print(row[1:], ',');

NameError: name 'K' is not defined

In [25]:
%%sql

DROP TABLE IF EXISTS sql_receipts;


CREATE TABLE sql_receipts (
receipt_id TEXT ,
description TEXT,
rewardsGroup TEXT,
bonusPointsEarned INTEGER ,
bonusPointsEarnedReason TEXT ,
createDate DATETIME,
dateScanned DATETIME,
finishedDate DATETIME,
modifyDate DATETIME,
pointsAwardedDate DATETIME,
pointsEarned TEXT,
purchaseDate DATETIME,
purchasedItemCount INTEGER,
rewardsReceiptStatus TEXT ,
totalSpent TEXT,
userId TEXT,
rewardsGroup_cleaned TEXT
    
);

 * sqlite:///engine.sqlite
Done.
Done.


[]

In [40]:
%%sql

SELECT * FROM sql_receipts
LIMIT 5;

 * sqlite:///engine.sqlite
Done.


[('5ff1e1eb0a720f0523000575', 'ITEM NOT FOUND', 'Null', 500, 'Receipt number 2 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)', '2021-01-03 15:25:31', '2021-01-03 15:25:31', '2021-01-03 15:25:31', '2021-01-03 15:25:36', '2021-01-03 15:25:31', '500.0', '2021-01-03 00:00:00', 5, 'FINISHED', '26.00', '5ff1e1eacfcf6c399c274ae6', 'Null'),
 ('5ff1e1bb0a720f052300056b', 'ITEM NOT FOUND', 'Null', 150, 'Receipt number 5 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)', '2021-01-03 15:24:43', '2021-01-03 15:24:43', '2021-01-03 15:24:43', '2021-01-03 15:24:48', '2021-01-03 15:24:43', '150.0', '2021-01-02 15:24:43', 2, 'FINISHED', '11.00', '5ff1e194b6a9d73a3a9f1052', 'Null'),
 ('5ff1e1bb0a720f052300056b', 'DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCED FAT BAG 1 OZ', 'DORITOS SPICY SWEET CHILI SINGLE SERVE', 150, 'Receipt number 5 completed, bonus point schedule DEFAULT (5cefdcacf3693e0b50e83a36)', '2021-01-03 15:24:43', '2021-01-03 15:24:43', '2021-01-03 15:24:43', '2021-01-03 15:24:48', '2021-01-03 15:24:43', '150.0', '2021-01-02 15:24:43', 2, 'FINISHED', '11.00', '5ff1e194b6a9d73a3a9f1052', 'DORITOS'),
 ('5ff1e1f10a720f052300057a', 'Null', 'Null', 5, 'All-receipts receipt bonus', '2021-01-03 15:25:37', '2021-01-03 15:25:37', 'Null', '2021-01-03 15:25:42', 'Null', '5', '2021-01-03 00:00:00', 1, 'REJECTED', '10.00', '5ff1e1f1cfcf6c399c274b0b', 'Null'),
 ('5ff1e1ee0a7214ada100056f', 'ITEM NOT FOUND', 'Null', 5, 'All-receipts receipt bonus', '2021-01-03 15:25:34', '2021-01-03 15:25:34', '2021-01-03 15:25:34', '2021-01-03 15:25:39', '2021-01-03 15:25:34', '5.0', '2021-01-03 00:00:00', 4, 'FINISHED', '28.00', '5ff1e1eacfcf6c399c274ae6', 'Null')]

### Brands

In [28]:
with gzip.open("brands.json.gz", "r") as f:
    data_brands = [json.loads(line) for line in f]

brand_id=[]
for a in range(len(data_brands)):
    if '_id' in data_brands[a]:
        brand_id.append( data_brands[a]['_id']['$oid'] )
    else:
        brand_id.append('Null')
        
        
barcode=[]
for a in range(len(data_brands)):
    if 'barcode' in data_brands[a]:
        barcode.append( data_brands[a]['barcode'])
    else:
        barcode.append('Null')

brandCode=[]
for a in range(len(data_brands)):
    if 'brandCode' in data_brands[a]:
        brandCode.append( data_brands[a]['brandCode'])
    else:
        brandCode.append('Null')
        
category=[]
for a in range(len(data_brands)):
    if 'category' in data_brands[a]:
        category.append( data_brands[a]['category'])
    else:
        category.append('Null')
        
        
categoryCode=[]
for a in range(len(data_brands)):
    if 'categoryCode' in data_brands[a]:
        categoryCode.append( data_brands[a]['categoryCode'])
    else:
        categoryCode.append('Null')
        
        
cpg=[]
for a in range(len(data_brands)):
    if 'cpg' in data_brands[a]:
        cpg.append( data_brands[0]['cpg']['$id']['$oid'])
    else:
        cpg.append('Null')
        
topBrand=[]
for a in range(len(data_brands)):
    if 'topBrand' in data_brands[a]:
        topBrand.append( data_brands[a]['topBrand'])
    else:
        topBrand.append('Null')
        
name=[]
for a in range(len(data_brands)):
    if 'name' in data_brands[a]:
        name.append( data_brands[a]['name'])
    else:
        name.append('Null')
        
df_brands = pd.DataFrame({'brand_id':brand_id, 'barcode':barcode, 'brandCode':brandCode,
            'category':category, 'categoryCode':categoryCode, 'cpg':cpg, 'topBrand':topBrand, 
             'name':name  })

def func_brand(s):
    if s=='Null' or s.split()[0:2]==['test', 'brand']:
        return 'Null'
    else:
        k=s.split()
        return k[0].upper()
    
df_brands['name_cleaned'] = df_brands['name'].apply(func_brand)


df_brands.head()

brand_id       barcode                      brandCode  \
0  601ac115be37ce2ead437551  511111019862                           Null   
1  601c5460be37ce2ead43755f  511111519928                      STARBUCKS   
2  601ac142be37ce2ead43755d  511111819905  TEST BRANDCODE @1612366146176   
3  601ac142be37ce2ead43755a  511111519874  TEST BRANDCODE @1612366146051   
4  601ac142be37ce2ead43755e  511111319917  TEST BRANDCODE @1612366146827   

         category      categoryCode                       cpg topBrand  \
0          Baking            BAKING  601ac114be37ce2ead437550    False   
1       Beverages         BEVERAGES  601ac114be37ce2ead437550    False   
2          Baking            BAKING  601ac114be37ce2ead437550    False   
3          Baking            BAKING  601ac114be37ce2ead437550    False   
4  Candy & Sweets  CANDY_AND_SWEETS  601ac114be37ce2ead437550    False   

                        name name_cleaned  
0  test brand @1612366101024         Null  
1                  Starbucks    STARBUCKS  
2  test brand @1612366146176         Null  
3  test brand @1612366146051         Null  
4  test brand @1612366146827         Null

In [29]:
engine2= create_engine('sqlite://', echo = False)

df_brands = df_brands.applymap(str)

df_brands.to_sql('new2', con = engine2)

K= engine2.execute("SELECT * FROM new2").fetchall()

K= [a[1:] for a in K]

In [2]:
for row in K:
    print(row[1:], ',');

NameError: name 'K' is not defined

In [31]:
%%sql

DROP TABLE IF EXISTS sql_brands;

CREATE TABLE sql_brands (
brand_id TEXT,
barcode TEXT ,
brandCode TEXT,
category TEXT,    
categoryCode TEXT ,
cpg TEXT ,
topBrand TEXT,    
name TEXT,
name_cleaned TEXT
);

 * sqlite:///engine.sqlite
Done.
Done.


[]

In [41]:
%%sql

SELECT * FROM sql_brands 
LIMIT 5;

 * sqlite:///engine.sqlite
Done.


[('601ac115be37ce2ead437551', '511111019862', 'Null', 'Baking', 'BAKING', '601ac114be37ce2ead437550', '0', 'test brand @1612366101024', 'Null'),
 ('601c5460be37ce2ead43755f', '511111519928', 'STARBUCKS', 'Beverages', 'BEVERAGES', '601ac114be37ce2ead437550', '0', 'Starbucks', 'Starbucks'),
 ('601ac142be37ce2ead43755d', '511111819905', 'TEST BRANDCODE @1612366146176', 'Baking', 'BAKING', '601ac114be37ce2ead437550', '0', 'test brand @1612366146176', 'Null'),
 ('601ac142be37ce2ead43755a', '511111519874', 'TEST BRANDCODE @1612366146051', 'Baking', 'BAKING', '601ac114be37ce2ead437550', '0', 'test brand @1612366146051', 'Null'),
 ('601ac142be37ce2ead43755e', '511111319917', 'TEST BRANDCODE @1612366146827', 'Candy & Sweets', 'CANDY_AND_SWEETS', '601ac114be37ce2ead437550', '0', 'test brand @1612366146827', 'Null')]

### Users

In [42]:
with gzip.open("users.json.gz", "r") as f:
    data_users = [json.loads(line) for line in f]

user_id=[]
for a in range(len(data_users)):
    if '_id' in data_users[a]:
        user_id.append( data_users[a]['_id']['$oid'] )
    else:
        user_id.append('Null')
        
state=[]
for a in range(len(data_users)):
    if 'state' in data_users[a]:
        state.append( data_users[a]['state'])
    else:
        state.append('Null')        

createdDate=[]
for a in range(len(data_users)):
    if 'createdDate' in data_users[a]:
        createdDate.append(  pd.to_datetime(data_users[a]['createdDate']['$date'] , unit='ms'))
    else:
        createdDate.append('Null')
        
lastLogin=[]
for a in range(len(data_users)):
    if 'lastLogin' in data_users[a]:
        lastLogin.append(   pd.to_datetime(data_users[a]['lastLogin']['$date'] , unit='ms'))
    else:
        lastLogin.append('Null')
        
role=[]
for a in range(len(data_users)):
    if 'role' in data_users[a]:
        role.append( data_users[a]['role'] )
    else:
        role.append('Null')
        
        
active=[]
for a in range(len(data_users)):
    if 'active' in data_users[a]:
        active.append( data_users[a]['active'] )
    else:
        active.append('Null')
        
df_users = pd.DataFrame({'user_id':user_id, 'state':state, 'createdDate':createdDate,
            'lastLogin':lastLogin, 'role':role, 'active':active})
df_users.head()

user_id state             createdDate  \
0  5ff1e194b6a9d73a3a9f1052    WI 2021-01-03 15:24:04.800   
1  5ff1e194b6a9d73a3a9f1052    WI 2021-01-03 15:24:04.800   
2  5ff1e194b6a9d73a3a9f1052    WI 2021-01-03 15:24:04.800   
3  5ff1e1eacfcf6c399c274ae6    WI 2021-01-03 15:25:30.554   
4  5ff1e194b6a9d73a3a9f1052    WI 2021-01-03 15:24:04.800   

                    lastLogin      role  active  
0  2021-01-03 15:25:37.858000  consumer    True  
1  2021-01-03 15:25:37.858000  consumer    True  
2  2021-01-03 15:25:37.858000  consumer    True  
3  2021-01-03 15:25:30.597000  consumer    True  
4  2021-01-03 15:25:37.858000  consumer    True

In [43]:
engine3 = create_engine('sqlite://', echo = False)

df_users = df_users.applymap(str)

df_users.to_sql('new3', con = engine3)

K= engine3.execute("SELECT * FROM new3").fetchall()

K= [a[1:] for a in K]

In [3]:
for row in K:
    print(row[1:], ',');

NameError: name 'K' is not defined

In [45]:
%%sql

DROP TABLE IF EXISTS sql_users;


CREATE TABLE sql_users(
user_id TEXT ,
state TEXT,
createdDate DATETIME,
lastLogin DATETIME ,
role TEXT ,
active TEXT)
;


 * sqlite:///engine.sqlite
Done.
Done.


[]

In [47]:
%%sql

SELECT * FROM  sql_users
LIMIT 5;


 * sqlite:///engine.sqlite
Done.


[('5ff1e194b6a9d73a3a9f1052', 'WI', '2021-01-03 15:24:04.800000', '2021-01-03 15:25:37.858000', 'consumer', 'True'),
 ('5ff1e194b6a9d73a3a9f1052', 'WI', '2021-01-03 15:24:04.800000', '2021-01-03 15:25:37.858000', 'consumer', 'True'),
 ('5ff1e194b6a9d73a3a9f1052', 'WI', '2021-01-03 15:24:04.800000', '2021-01-03 15:25:37.858000', 'consumer', 'True'),
 ('5ff1e1eacfcf6c399c274ae6', 'WI', '2021-01-03 15:25:30.554000', '2021-01-03 15:25:30.597000', 'consumer', 'True'),
 ('5ff1e194b6a9d73a3a9f1052', 'WI', '2021-01-03 15:24:04.800000', '2021-01-03 15:25:37.858000', 'consumer', 'True')]

## Second: Write a query that directly answers a predetermined question from a business stakeholder

### What are the top 5 brands by receipts scanned for most recent month?

We assume most recent date (current date) is'2021-03-01', based on the Max datescanned of receipts

In [49]:
%%sql
SELECT MAX(dateScanned) FROM sql_receipts

 * sqlite:///engine.sqlite
Done.


[('2021-03-01 23:17:34.772000',)]

In [51]:
%%sql
WITH final AS(
WITH grouped AS
    (WITH filtered AS
        (SELECT * FROM sql_receipts
        WHERE DATE(dateScanned) BETWEEN '2021-02-01' AND '2021-02-28' AND rewardsGroup_cleaned!='Null')
    SELECT rewardsGroup_cleaned, COUNT(*) FROM filtered
    GROUP BY rewardsGroup_cleaned
    ORDER BY COUNT(*) DESC
    LIMIT 5)
    
SELECT grouped.rewardsGroup_cleaned , sql_brands.name_cleaned, sql_brands.name  FROM grouped
LEFT JOIN sql_brands
ON UPPER(rewardsGroup_cleaned) = UPPER(name_cleaned))

    
SELECT 
CASE
WHEN name IS NOT NULL THEN UPPER(name)
ELSE UPPER(rewardsGroup_cleaned)
END AS Brand_Name
FROM final

 * sqlite:///engine.sqlite
Done.


[('SARGENTO',), ('SWISS',), ('HEINZ',), ('DUNCAN',), ('DORITOS',)]

we can see names might be null but we extract name from rewardsGroup in the rewardsReceiptItemList column

we can see the 5 top brands are SARGENTO, SWISS, HEINZ, DUNCAN, DORITOS

### How does the ranking of the top 5 brands by receipts scanned for the recent month compare to the ranking for the previous month?

In [52]:
%%sql
WITH result AS(WITH final AS(

WITH grouped AS
    (WITH filtered AS
        (SELECT * FROM sql_receipts
        WHERE DATE(dateScanned) BETWEEN '2021-01-01' AND '2021-01-31' AND rewardsGroup_cleaned!='Null')
    SELECT rewardsGroup_cleaned, COUNT(*) FROM filtered
    GROUP BY rewardsGroup_cleaned
    ORDER BY COUNT(*) DESC
    LIMIT 5)
    
SELECT grouped.rewardsGroup_cleaned , sql_brands.name_cleaned, sql_brands.name  FROM grouped
LEFT JOIN sql_brands
ON UPPER(rewardsGroup_cleaned) = UPPER(name_cleaned))

SELECT 
CASE
WHEN  name IS NOT NULL THEN UPPER(name)
ELSE UPPER(rewardsGroup_cleaned)
END AS Brand_Name
FROM final)
SELECT DISTINCT(Brand_Name) FROM result

 * sqlite:///engine.sqlite
Done.


[("BEN & JERRY'S",),
 ('SARGENTO',),
 ('HUGGIES LITTLE SWIMMERS',),
 ('HUGGIES',),
 ('KNORR',),
 ('PEPSI DIET',),
 ('PEPSI',),
 ('PEPSI MAX',)]

I we assume that for example HUGGIES LITTLE SWIMMERS is different from HUGGIES  or PEPSI DIET is different from PEPSI
and PEPSI MAX. Based on this assumption,  the top 5 brands for the previous month (BETWEEN '2021-01-01' AND '2021-01-31') are BEN & JERRY'S, SARGENTO, HUGGIES LITTLE SWIMMERS, HUGGIES and KNORR.

However, if we assume that HUGGIES LITTLE SWIMMERS is the same as  HUGGIES  or PEPSI DIET is the same as PEPSI
and PEPSI MAX, the top 5 brands for the previous month (BETWEEN '2021-01-01' AND '2021-01-31') would be BEN & JERRY'S, SARGENTO, HUGGIES, KNORR and PEPSI.

 
On the other hand, for the most recent month, the the top 5 brands were SARGENTO, SWISS, HEINZ, DUNCAN and DORITOS.

###  When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

I assumed that finished means accepted

In [53]:
%%sql
SELECT rewardsReceiptStatus, ROUND(AVG(totalSpent)) AS Average_Spend FROM sql_receipts
WHERE rewardsReceiptStatus='FINISHED'
GROUP BY rewardsReceiptStatus
UNION
SELECT rewardsReceiptStatus, ROUND(AVG(totalSpent)) AS Average_Spend FROM sql_receipts
WHERE rewardsReceiptStatus='REJECTED'
GROUP BY rewardsReceiptStatus

 * sqlite:///engine.sqlite
Done.


[('FINISHED', 1244.0), ('REJECTED', 20.0)]

The above table shows that average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ (Finished) is greater

### When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [412]:
%%sql
SELECT rewardsReceiptStatus, SUM(purchasedItemCount) AS Total_Number_Purchased FROM sql_receipts
WHERE rewardsReceiptStatus='FINISHED'
GROUP BY rewardsReceiptStatus
UNION
SELECT rewardsReceiptStatus, SUM(purchasedItemCount) AS Total_Number_Purchased FROM sql_receipts
WHERE rewardsReceiptStatus='REJECTED'
GROUP BY rewardsReceiptStatus

   sqlite:///engine.sqlite
 * sqlite:///save_pandas.db.sqlite
   sqlite:///testdb.sqlite
Done.


[('FINISHED', 1364998), ('REJECTED', 740)]

The above table shows that total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ (Finished)
is greater?

### Which brand has the most spend among users who were created within the past 6 months?

Based on Max date of receipts, we assume most recent date (current date) is'2021-03-01' 

In [480]:
%%sql
WITH result AS(
WITH final AS(
WITH merged AS( WITH tab AS(
SELECT * FROM  sql_users
WHERE DATE(createdDate) BETWEEN   '2020-09-01' AND '2021-03-01')

SELECT tab.user_id, totalSpent, rewardsGroup_cleaned FROM tab
INNER JOIN sql_receipts
ON tab.user_id= sql_receipts.userId
WHERE rewardsGroup_cleaned!='Null' )

SELECT  sql_brands.name,name_cleaned, rewardsGroup_cleaned, totalSpent  FROM merged
LEFT JOIN sql_brands
ON UPPER(rewardsGroup_cleaned) = UPPER(name_cleaned))

SELECT 
CASE
WHEN  name IS NOT NULL THEN UPPER(name)
ELSE UPPER(rewardsGroup_cleaned)
END AS Brand_Name,
totalSpent
FROM final)
SELECT Brand_Name, SUM( totalSpent) AS Sum_Total_Spent FROM result
GROUP BY Brand_Name
ORDER BY SUM( totalSpent) DESC
LIMIT 1

   sqlite:///engine.sqlite
 * sqlite:///save_pandas.db.sqlite
   sqlite:///testdb.sqlite
Done.


[('HUGGIES', 256199.2)]

Huggies has the most spend among users who were created within the past 6 months, assuming that the current date is 2021-03-01

### Which brand has the most transactions among users who were created within the past 6 months?

In [497]:
%%sql
WITH result AS(
WITH final AS(
WITH merged AS(WITH tab AS(
SELECT * FROM  sql_users
WHERE DATE(createdDate) BETWEEN   '2020-09-01' AND '2021-03-01')

SELECT receipt_id,  rewardsGroup_cleaned FROM tab
INNER JOIN sql_receipts
ON tab.user_id= sql_receipts.userId
WHERE rewardsGroup_cleaned!='Null')

SELECT  sql_brands.name, name_cleaned, rewardsGroup_cleaned, receipt_id  FROM merged
LEFT JOIN sql_brands
ON UPPER(rewardsGroup_cleaned) = UPPER(name_cleaned))

SELECT 
CASE
WHEN  name IS NOT NULL THEN UPPER(name)
ELSE UPPER(rewardsGroup_cleaned)
END AS Brand_Name,
receipt_id
FROM final)

SELECT Brand_Name, COUNT(receipt_id ) AS Number_Transactions FROM result
GROUP BY Brand_Name
ORDER BY COUNT(receipt_id) DESC
LIMIT 1

   sqlite:///engine.sqlite
 * sqlite:///save_pandas.db.sqlite
   sqlite:///testdb.sqlite
Done.


[('HUGGIES', 3080)]

Huggies has the most number of transactions among users who were created within the past 6 months, assuming that 
the current date is 2021-03-01

## Third: Evaluate Data Quality Issues in the Data Provided

First I checked the percentage of **Missing Values**  in different columns in 3 dataframes, including df_receipts,df_brands and df_users.

In [150]:
val=[]
ind=[]
for a in df_receipts.columns[:-1]:
    ind.append(a)
    val.append(100*(df_receipts[df_receipts[a]=='Null'].shape[0])/(df_receipts[a].shape[0]))
pd.Series(data= val, index=ind).sort_values(ascending=False)

rewardsGroup               76.547893
finishedDate               19.116651
bonusPointsEarnedReason    18.981168
bonusPointsEarned          18.981168
pointsAwardedDate          17.626338
pointsEarned               15.282482
description                11.123154
purchasedItemCount          6.557377
purchaseDate                6.205121
totalSpent                  5.893510
userId                      0.000000
rewardsReceiptStatus        0.000000
modifyDate                  0.000000
dateScanned                 0.000000
createDate                  0.000000
receipt_id                  0.000000
dtype: float64

We can see that 76.5% of data in the rewardsGroup column extracted from rewardsReceiptItemList column in df_receipt is missing. Also, other columns have relatively high percentage of missing values.            

In [153]:
val=[]
ind=[]
for a in df_brands.columns[:-1]:
    ind.append(a)
    val.append(100*(df_brands[df_brands[a]=='Null'].shape[0])/(df_brands[a].shape[0]))
pd.Series(data= val, index=ind).sort_values(ascending=False)

categoryCode    55.698372
topBrand        52.442159
brandCode       20.051414
category        13.281919
name             0.000000
cpg              0.000000
barcode          0.000000
brand_id         0.000000
dtype: float64

We can see that about 55.7% of data in the categoryCode column in df_brands is missing. Also, other columns such as topBrand have relatively high percentage of missing values.

In [152]:
val=[]
ind=[]
for a in df_users.columns[:-1]:
    ind.append(a)
    val.append(100*(df_users[df_users[a]=='Null'].shape[0])/(df_users[a].shape[0]))
pd.Series(data= val, index=ind).sort_values(ascending=False)

lastLogin      12.525253
state          11.313131
role            0.000000
createdDate     0.000000
user_id         0.000000
dtype: float64

We can see that about 12.5% of data in the lastLogin column in df_users is missing. Also,  11.3% data in the state column in missing.

Now I check the percentage of **Outliers** for one of the columns (totalSpent) in df_receipts.

In [176]:
Q_25 = df_receipts[df_receipts['totalSpent']!='Null']['totalSpent'].apply(lambda x: float(x)).quantile(q=0.25)
Q_75 = df_receipts[df_receipts['totalSpent']!='Null']['totalSpent'].apply(lambda x: float(x)).quantile(q=0.75)

res=[]
for a in df_receipts[df_receipts['totalSpent']!='Null']['totalSpent'].apply(lambda x: float(x)):
    if (a>Q_75+1.5*IQR or a<Q_25-1.5*IQR) and (a not in res):
        res.append(a)
        
df_receipts_without_null_totalspent = df_receipts[df_receipts['totalSpent']!='Null']
df_receipts_without_null_totalspent['totalSpent'] = df_receipts_without_null_totalspent['totalSpent'].apply(lambda x: float(x))

Percentage_of_outliers = 100*df_receipts_without_null_totalspent[df_receipts_without_null_totalspent['totalSpent'].isin(res)].shape[0]/df_receipts_without_null_totalspent['totalSpent'].shape[0]
print('Percentage_of_outliers in total_spent column is', round(Percentage_of_outliers, 2))

Percentage_of_outliers in total_spent column is 18.57


In [ ]:
We can see that the percentage_of_outliers in total_spent column is 18.57% which is relatively hight and should be removed.

## Third: Communicate with Stakeholders

Dear Sir,


I analyzed the data as per your request and you can find the results as attached. There are some issues I would like to mention here.

As it is very important to make a relationship between the receipts data and brands data, the quality of data was evaluated. However, there is not an easily detectable relationship between these two data. Therefore, I extracted the rewardsGroup data from rewardsReceiptItemList in receipt table. This data contained the name of brands, so I related it to brand names in brands table. However, there are some inconsistencies between the rewardsGroup data and brands name.


It is noteworthy to mention that rewardsReceiptItemList in receipt table contains barcode data which should be consistent with the barcode data in brands table, but the range of barcodes in brands table and receipt table are totally different. Could you please clarify if the barcodes in receipt table are valid? Do you have any preference for selecting the column that can be used making a relationship between the receipt table and brands table?


Regarding the quality issues, the percentage of outliers in total spent column in receipts table is relatively high, so the outliers should be removed. Moreover, we can see that 76.5% of data in the rewardsGroup column extracted from rewardsReceiptItemList column in receipt table is missing. Also, other columns of this table have relatively high percentage of missing values.  In the brands table, 55.7% of data in the categoryCode column is missing. Also, other columns such as topBrand have relatively high percentage of missing values. In the users table, two columns have slightly high percentage of missing values.


I order to resolve the quality issues, we can filter out the rows in the columns in which the percentage of missing values are very small and less than oner percent. Also, we can drop the columns that have more than 70% missing values. For the columns that have less than 70% missing values we can fill in the missing values with mean or median of the column.  We can also fill these missing values by Imputation using deep learning or even predict the missing values Using random forests algorithm.
If you gather and provide some additional information, we can optimize the data assets. For example, if we have the age and gender of users, we can categorize the brands with the highest total spent for men and women or categorize them based on the users’ age. Moreover, if it is possible to provide the missing data in each column for receipts, brands and users data, a more accurate analysis could be performed.


Finally, in order to scale the production, you can use a more robust caching engine, scale your server, ease server load or reduce read load by adding more read replicas.


Please let me know if you need any further information. I look forward to hearing about you.



Kind regards,


Soha
